### PR TITLE
M0-1-3: Exclude uninstantiated variable templates

### LINKED ISSUE #619: `M0-1-3`: Exclude uninstantiated variable templates
### Affected rules
 - `M0-1-3`

### Description

Exclude variables from uninstantiated templates and compiler generated variables from this query.

### Example

```cpp
template <bool... Args>
extern constexpr bool all_of_v = true;

template <bool B1, bool... Args>
extern constexpr bool all_of_v<B1, Args...> = B1 && all_of_v<Args...>;

void test() {
  all_of_v<true,true,true>;
}
```

### PR DESCRIPTION
**Pull request by lcartey**
## Description

Fixes #619.

Note: I have also chosen to exclude compiler generated variables for `UnusedGlobalOrNamespaceVariable.ql`, and both uninstantiated templates and compiler generated variables for the other `M0-1-3` queries. However, I do not have any specific examples where these would make a difference for these other cases, so I have chosen not to include them in the change note.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `M0-1-3`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Review by knewbury01 — CHANGES_REQUESTED** [review]
other than few formatting/polish items, looks good, would be happy to approve after these are addressed!

File: cpp/common/src/codingstandards/cpp/deadcode/UnusedVariables.qll [hunk 1]
```diff
@@ -49,6 +49,11 @@ class PotentiallyUnusedLocalVariable extends LocalVariable {
       // Ignore functions with error expressions as they indicate expressions that the extractor couldn't process
       not any(ErrorExpr e).getEnclosingFunction() = f
     )
+    // ) and
```
[the remark this answers is not in the document]
Comment (lcartey) [reply cpp/common/src/codingstandards/cpp/deadcode/UnusedVariables.qll:52]: Yes, this was not intended.

File: cpp/autosar/test/rules/M0-1-3/UnusedLocalVariable.expected [hunk 2]
```diff
@@ -4,3 +4,4 @@
 | test.cpp:23:5:23:5 | t | Local variable 't' in 'f1' is not used. |
```
[the remark this answers is not in the document]
Comment (lcartey) [reply cpp/autosar/test/rules/M0-1-3/UnusedLocalVariable.expected:4]: Ah, forgot to commit it. Fixed.

**Review by knewbury01 — APPROVED** [review]
lgtm! thanks for the changes 👍

**Pull request merged**